File: strings/string_permutations.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm> 
using namespace std;

//  swap values for x and y
void swap(char &x, char &y) {
    char temp;
    temp = x;
    x = y;
    y = temp;
}

//print permutations of string using recursion with duplicates
void permute_recursion_with_duplicates(string a, int l_index, int r_index) {
    if (l_index == r_index)
        cout << a << endl;
    else {
        for (int i = l_index; i <= r_index; i++) {
            swap( a[l_index], a[i]);
            permute_recursion_with_duplicates(a, l_index+1, r_index);
            swap( a[l_index], a[i]);
        }
    }
}

//print permutations of string using recursion without duplicates
void permute_recursion_without_duplicates(string a, int l_index, int r_index) {
    if (l_index == r_index)
        cout << a << endl;
    else {
        for (int i = l_index; i <= r_index; i++) {
            if((i == l_index) || (a[i] != a[i-1])) {
                swap( a[l_index], a[i]);
                permute_recursion_without_duplicates(a, l_index+1, r_index);
                swap( a[l_index], a[i]);
            }
        }
    }
}

int main() {
    // Using recursion
    // Time complexity - O(n*n!)
    // requires O(n) time to print a permutation, total n! permutations
    string str = "BAA";
    int length = str.length();
    // print permutations in lexicographical order
    std::sort(str.begin(), str.end());

    // prints duplicates
    permute_recursion_with_duplicates(str, 0, length-1);
    cout << endl;

    // no duplicates
    permute_recursion_without_duplicates(str, 0, length-1);
    cout << endl;

    return 0;
}

File: strings/string_permutations.py
"""
Print all permutatinos of string

E.g.
For "BAA"
AAB
ABA
BAA
"""

# -*- encoding: utf-8 *-*


def swap(char1, char2):
    print(char1, " ", char2)
    temp = char1
    char1 = char2
    char2 = temp
    print(char1, " ", char2)
    # input = input[:left_index + i] + input[right_index - i] + input[left_index + i + 1:]
    return


def permute_recursion_with_duplicates(input, l_index, r_index):
    """
    """
    if l_index == r_index:
        print(input)
    else:
        for i in range(l_index, r_index):
            swap(input[l_index], input[i])
            permute_recursion_with_duplicates(input, l_index + 1, r_index)
            swap(input[l_index], input[i])


def permute_recursion_without_duplicates(input, l_index, r_index):
    """
    """
    return


def main():
    input = "AB"

    # print permutations in lexicographical order
    # sort(str.begin(), str.end());

    permute_recursion_with_duplicates(input, 0, len(input))
    permute_recursion_without_duplicates(input, 0, len(input))
    return


if __name__ == '__main__':
    main()

